File: core/RestoreIpAddresses/RestoreIpAddresses.java
package com.dbc;

import java.util.*;

public class RestoreIpAddresses {
    private final Set<String> remain = new HashSet<>();
    private final List<String> res = new ArrayList<>();

    private void dfs(String subStr, String prefix, int num) {
        if (num == 4 && subStr.length() > 0 && subStr.length() <= 3 && Integer.parseInt(subStr) >= 0 &&
                Integer.parseInt(subStr) <= 255) {
            if (subStr.startsWith("0") && subStr.length() > 1) return;
            prefix = prefix + "." + subStr;
            if (!this.remain.contains(prefix)) {
                this.res.add(prefix);
                this.remain.add(prefix);
            }
        }
        if (num < 4) {
            for (int i = 1; i < Math.min(4, subStr.length()); i++) {
                String tempStr = subStr.substring(0, i);
                if ((tempStr.startsWith("0") && tempStr.length() > 1) || Integer.parseInt(tempStr) < 0 || Integer.parseInt(tempStr) > 255) continue;
                if (num != 1) tempStr = prefix + "." + tempStr;
                dfs(subStr.substring(i), tempStr, num + 1);
            }
        }
    }

    public List<String> restoreIpAddresses(String s) {
        dfs(s, "", 1);
        return this.res;
    }
}

File: core/RestoreIpAddresses/RestoreIpAddresses.py
from typing import List

def restoreIpAddresses(s: str) -> List[str]:
    res, remain = [], {}
    def dfs(sub_str, num, sub_res):
        if num == 4 and 0 < len(sub_str) <= 3 and 0 <= int(sub_str) <= 255:
            if len(sub_str) > 1 and sub_str.startswith("0"):
                return
            sub_res.append(sub_str)
            temp = ".".join(sub_res)
            if temp not in remain:
                res.append(temp)
                remain[temp] = True
            sub_res.pop()
        if num < 4:
            for i in range(1, 4):
                if (sub_str.startswith("0") and i != 1) or sub_str[:i] == "" or not (0 <= int(sub_str[:i]) <= 255):
                    continue
                sub_res.append(sub_str[:i])
                dfs(sub_str[i:], num + 1, sub_res)
                sub_res.pop()

    dfs(s, 1, [])
    return res

if __name__ == '__main__':
    print(restoreIpAddresses("0000"))



